qcengine/procedures/berny.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+import traceback
+from typing import Any, Dict, Union
+import pkg_resources
+
+import qcengine
+from qcelemental.models import OptimizationInput, OptimizationResult
+from qcelemental.util import which_import
+
+from .model import ProcedureHarness
+from ..config import TaskConfig
+
+# TODO there is a lot of boilerplate shared with GeometricProcedure
+# TODO how to handle logging?
+# TODO where to store paramters for Berny?
+
+
+class BernyProcedure(ProcedureHarness):
+
+    _defaults = {"name": "Berny", "procedure": "optimization"}
+
+    class Config(ProcedureHarness.Config):
+        pass
+
+    def found(self, raise_error: bool = False) -> bool:
+        return which_import(
+            "berny", return_bool=True, raise_error=raise_error, raise_msg="Please install via `pip install pyberny`.",
+        )
+
+    def build_input_model(self, data: Union[Dict[str, Any], "OptimizationInput"]) -> "OptimizationInput":
+        return self._build_model(data, OptimizationInput)
+
+    def compute(self, input_data: "OptimizationInput", config: "TaskConfig") -> "OptimizationResult":
+        try:
+            import berny
+        except ModuleNotFoundError:
+            raise ModuleNotFoundError("Could not find Berny in the Python path.")
+        input_data = input_data.dict()
+        geom_qcng = input_data["initial_molecule"]
+        comput = {**input_data["input_specification"], "molecule": geom_qcng}
+        trajectory = []
+        output_data = input_data.copy()
+        try:
+            geom_berny = berny.Geometry(geom_qcng["symbols"], geom_qcng["geometry"])
```

Comment:
Do you use atomic units for everything internally?

Reply:
Yes.

Reply:
x

Reply:
Erm, clearly I should be reading my own documentation. Pyberny works in angstroms, but so does QCEngine, so everything checks out?

Reply:
QCEngine woks exclusively in Bohr.

Reply:
But `Molecule`'s API uses angstroms, or? At least I thought so based on this paragraph: http://docs.qcarchive.molssi.org/projects/QCElemental/en/stable/model_molecule.html#fragments

Reply:
Ah, the _string initialization_ of `Molecule` defaults to angstroms (you can add a `units au` line to the string to make it interpret as Bohr). But the internal storage is stringently au. By the way, there are array or dictionary initializers, too, if something like that would be more convenient.

You can consult http://docs.qcarchive.molssi.org/projects/QCElemental/en/stable/model_molecule.html#fields for the internal fields. (Ignore that some have trailing underscore (e.g., `masses_) -- that's a detail that allows skipping storage of fields with default value.)

Reply:
Interesting! So indeed this was incorrect and had to add the conversion factor between angstrom and bohr. What's interesting is that the optimization worked anyway, albeit less efficiently.

As long as the coordinates/gradients are scaled contra/covariantly (as was the case here), the gradient descent is unaffected. The only difference was that Pyberny detected the molecule as three unbonded atoms, that is, as a van der Waals cluster—which I should have notice. But it sill optimized it.

Anyway, thanks for the catch.